va/vampic.go

```diff
@@ -0,0 +1,433 @@
+				return
+			}
+			responses <- &response{rva.Address, rva.Perspective, rva.RIR, res, err}
+		}(va.remoteVAs[i])
+	}
+
+	required := remoteVACount - va.maxRemoteFailures
+	var passed []string
+	var failed []string
+	var passedRIRs = map[string]struct{}{}
+	var firstProb *probs.ProblemDetails
+
+	for resp := range responses {
+		var currProb *probs.ProblemDetails
+
+		if resp.err != nil {
+			// Failed to communicate with the remote VA.
+			failed = append(failed, resp.perspective)
+
+			if core.IsCanceled(resp.err) {
+				currProb = probs.ServerInternal("Secondary validation RPC canceled")
+			} else {
+				va.log.Errf("Operation on remote VA (%s) failed: %s", resp.addr, resp.err)
+				currProb = probs.ServerInternal("Secondary validation RPC failed")
+			}
+		} else if resp.result.GetProblem() != nil {
+			// The remote VA returned a problem.
+			failed = append(failed, resp.perspective)
+
+			var err error
+			currProb, err = bgrpc.PBToProblemDetails(resp.result.GetProblem())
+			if err != nil {
+				va.log.Errf("Operation on Remote VA (%s) returned malformed problem: %s", resp.addr, err)
+				currProb = probs.ServerInternal("Secondary validation RPC returned malformed result")
+			}
+		} else {
+			// The remote VA returned a successful result.
+			passed = append(passed, resp.perspective)
+			passedRIRs[resp.rir] = struct{}{}
+		}
+
+		if firstProb == nil && currProb != nil {
+			// A problem was encountered for the first time.
+			firstProb = currProb
+		}
+
+		// To respond faster, if we get enough successes or too many failures, we cancel remaining RPCs.
+		// Finish the loop to collect remaining responses into `failed` so we can rely on having a response
+		// for every request we made.
+		if len(passed) >= required && len(passedRIRs) >= requiredRIRs {
+			cancel()
+		}
+		if len(failed) > va.maxRemoteFailures {
+			cancel()
+		}
+
+		// Once all the VAs have returned a result, break the loop.
+		if len(passed)+len(failed) >= remoteVACount {
+			break
+		}
+	}
+	if len(passed) >= required && len(passedRIRs) >= requiredRIRs {
+		return summarizeMPIC(passed, failed, passedRIRs), nil
+	} else if len(passed) >= required && firstProb == nil {
+		// This should never happen. If we didn't meet the thresholds above we
+		// should have seen at least one error.
+		return summarizeMPIC(passed, failed, passedRIRs), probs.Unauthorized(
+			"During secondary validation: validation could not be corroborated by enough distinct RIRs",
+		)
+	} else if len(failed) > va.maxRemoteFailures {
+		firstProb.Detail = fmt.Sprintf("During secondary validation: %s", firstProb.Detail)
+		return summarizeMPIC(passed, failed, passedRIRs), firstProb
+	} else {
+		// This condition should not occur - it indicates the passed/failed counts
+		// neither met the required threshold nor the maxRemoteFailures threshold.
+		return summarizeMPIC(passed, failed, passedRIRs), probs.ServerInternal("Too few remote RPC results")
+	}
```

Comment:
I think checking for `&& firstProb == nil` and having the second case say "this should never happen" is confusing. I struggle to answer the question "which case do we hit if we get enough passed responses but they're not from enough different RIRs?"

I think that this confusion can be fixed by essentially swapping the order of the second and third cases, as I mentioned at the bottom of my previous comment.

Then we would have four clear cases, in order:
- enough passed, from enough RIRs (happy path, return success)
- too many failed (return firstProb)
- too few RIRs (return custom error)
- else (this condition should not occur)